From d99764eaa9dc816d0f7363c1f465daf4705936a6 Mon Sep 17 00:00:00 2001
From: equiroz <equiroz@ciencias.unam.mx>
Date: Sun, 16 Feb 2020 17:59:26 -0600
Subject: [PATCH 2/3] Adding thread comparaison funciont implementation

---
 lab/pintos/src/threads/thread.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/lab/pintos/src/threads/thread.c b/lab/pintos/src/threads/thread.c
index 955ccdc..2cec2fc 100644
--- a/lab/pintos/src/threads/thread.c
+++ b/lab/pintos/src/threads/thread.c
@@ -322,6 +322,17 @@ thread_yield (void)
   intr_set_level (old_level);
 }
 
+/* EXTRA1: Implementing comparaison function between threads based on their time 
+left to sleep*/
+bool 
+thread_less_func(const struct list_elem *a, const struct list_elem *b, void *aux)
+{
+  struct thread *ta = list_entry(a, struct thread, allelem);
+  struct thread *tb = list_entry(b, struct thread, allelem);
+  
+  return ta->ticks_alarm < tb->ticks_alarm;
+}
+
 /* Invoke function 'func' on all threads, passing along 'aux'.
    This function must be called with interrupts off. */
 void
-- 
2.25.0

