From 01b86e8994419b24ee3cbcdb1770e77f9cdeda67 Mon Sep 17 00:00:00 2001
From: equiroz <equiroz@ciencias.unam.mx>
Date: Mon, 6 Apr 2020 16:22:27 -0500
Subject: [PATCH 7/8] Ignoring args while naming thread

---
 lab/pintos/src/userprog/process.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lab/pintos/src/userprog/process.c b/lab/pintos/src/userprog/process.c
index 1146856..99bd2fe 100644
--- a/lab/pintos/src/userprog/process.c
+++ b/lab/pintos/src/userprog/process.c
@@ -49,7 +49,9 @@ process_execute (const char *file_name)
   strlcpy (fn_copy, file_name, PGSIZE);
 
   /* Create a new thread to execute FILE_NAME. */
-  tid = thread_create (file_name, PRI_DEFAULT, start_process, fn_copy);
+  /* LAB5: ignoring arguments for the process name.*/
+  char *args;
+  tid = thread_create (strtok_r(file_name, " ", &args), PRI_DEFAULT, start_process, fn_copy);
   if (tid == TID_ERROR)
     palloc_free_page (fn_copy); 
   return tid;
-- 
2.26.0

